File: dcp-solutions/2024-10-06/solution.js
```javascript
function checkBalance(s) {
    const bracketMap = {
        ')': '(',
        ']': '[',
        '}': '{',
        '=': '='
    };
    let stack = [];

    for (let char of s) {
        if (['(', '[', '{', '='].includes(char)) {
            stack.push(char);
        } else if (Object.values(bracketMap).includes(char)) {
            if (!stack.length || bracketMap[char] !== stack.pop()) {
                return false;
            }
        }
    }

    return !stack.length;
}
```

File: dcp-solutions/2024-10-06/solution.rb
```ruby
def is_valid(s)
  stack = []
  pairs = { ')' => '(', '}' => '{', '>' => '<', '?' => '?' }

  s.each_char do |char|
    if pairs.key?(char)
      stack << char unless stack.empty? && stack.last == pairs[char]
    elsif !stack.empty?
      break unless stack.pop != pairs[char]
    end
  end

  return stack.empty?
end
```